File: chap1/sec1.5/prob1.7/sol_integration.m
% Settings for Matlab
close all; clear;
set( 0, 'defaultAxesTickLabelInterpreter', 'latex' );
set( 0, 'defaultLegendInterpreter',        'latex' );
set( 0, 'defaultTextInterpreter',          'latex' );
set( 0, 'defaultAxesFontSize', 12 );
format long 

% Parameters
nStep = 10000;
T = 1/2;
method = 'FE';

% Parameters for numerical solution
L = 1;
J = 99;
A = FinDiffMatrixD2C2(J, L);
x = linspace(0, 1, J+2);
x = x(2:end-1);

% Parameters for analytical solution
nF = 5;
n = 0:nF-1;
global nMesh xMesh;
[xMesh, nMesh] = meshgrid(x, n);

u0 = 20*ones(J, 1);

if strcmp(method, 'FE')
    [t, u] = ForwardEulerLin(A, u0, T, nStep);
elseif strcmp(method, 'BE')
    [t, u] = BackwardEulerLin(A, u0, T, nStep);
end

plot(x, u(:, int64(nStep/4)+1), 'DisplayName', '$T=1/8$');
hold on;
plot(x, u(:, int64(nStep/2)+1), 'DisplayName', '$T=1/4$');
plot(x, u(:, nStep+1), 'DisplayName', '$T=1/2$');
gray = [0.5,0.5,0.5];
plot(x, uTh(1/8), 'o', 'Color', gray, 'HandleVisibility', 'off', ...
    'MarkerIndices',1:5:length(x));
plot(x, uTh(1/4), 'o', 'Color', gray, 'HandleVisibility', 'off', ...
    'MarkerIndices',1:5:length(x))
plot(x, uTh(1/2), 'o', 'Color', gray, 'HandleVisibility', 'off', ...
    'MarkerIndices',1:5:length(x))
xlabel('$x$'); legend(); grid on;

% Function to compute the analytical solution at time t
function sol=uTh(t)
    global nMesh xMesh; 
    s = (80/pi)./(2*nMesh+1) .* exp(-(2*nMesh+1).^2 * pi^2 * t) ...
        .* sin((2*nMesh+1).*pi .*xMesh);
    sol = sum(s, 1);
end

File: chap1/sec1.5/prob1.7/sol_integration.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Wed Sep 19 00:46:35 2018

@author: telu
"""
import numpy as np
import matplotlib.pyplot as plt
from solver import forwardEulerLin, backwardEulerLin

N = 64
T = 1/2
method = 'BE'

J = 99
h = 1/(J+1)
A = np.diag(np.ones(J-1), k=1) \
    + np.diag(np.ones(J-1), k=-1) \
    - 2*np.diag(np.ones(J), k=0)
A /= h**2

x = np.linspace(0, 1, num=J+2)[1:-1]

u0 = np.ones(J)
u0 *= 20

if method == 'FE':
    t, u = forwardEulerLin(A, u0, T, N)
elif method == 'BE':
    t, u = backwardEulerLin(A, u0, T, N)


fig = plt.figure('Solution')
plt.plot(x, u[:, N//4], '-', label='$T=1/8$', markevery=0.05)
plt.plot(x, u[:, N//2], '-', label='$T=1/4$', markevery=0.05)
plt.plot(x, u[:, N], '-', label='$T=1/2$', markevery=0.05)

nF = 5
pi = np.pi
n = np.arange(nF)
xTh = np.linspace(0, 1, num=500)
xMesh, nMesh = np.meshgrid(xTh, n)


def uTh(t):
    s = 80/pi/(2*nMesh+1) * np.exp(-(2*nMesh+1)**2 * pi**2 * t) \
        * np.sin((2*nMesh+1)*pi*xMesh)
    return np.sum(s, axis=0)


plt.plot(xTh, uTh(1/8), 'o', c='gray', markevery=0.05)
plt.plot(xTh, uTh(1/4), 'o', c='gray', markevery=0.05)
plt.plot(xTh, uTh(1/2), 'o', c='gray', markevery=0.05)

plt.xlabel('$x$')
plt.ylabel('$u(t)$')
plt.legend()
plt.grid(True)
fig.set_tight_layout(True)
plt.show()
